lib/models.js

```diff
@@ -0,0 +1,223 @@
+
+        // Set non-writable properties.
+        defineImmutable(this, 'directory', directory);
+        defineImmutable(this, 'fragment', fragment);
+        defineImmutable(this, 'full', directory + fragment);
+        defineImmutable(this, 'sep', sep);
+    }
+
+    @cachedProperty
+    get stat() {
+        try {
+            return fs.statSync(absolutify(this.full));
+        } catch (err) {
+            if (err.code === 'ENOENT') {
+                return null;
+            } else {
+                throw err;
+            }
+        }
+    }
+
+    isDirectory() {
+        return this.stat ? this.stat.isDirectory() : null;
+    }
+
+    isFile() {
+        return this.stat ? !this.stat.isDirectory() : null;
+    }
+
+    isProjectDirectory() {
+        return atom.project.getPaths().indexOf(this.full) !== -1;
+    }
+
+    isRoot() {
+        return stdPath.dirname(this.full) === this.full;
+    }
+
+    hasCaseSensitiveFragment() {
+        return this.fragment !== '' && this.fragment !== this.fragment.toLowerCase();
+    }
+
+    exists() {
+        return this.stat !== null;
+    }
+
+    asDirectory() {
+        return new Path(this.full + (this.fragment ? this.sep : ''));
+    }
+
+    parent() {
+        if (this.isRoot()) {
+            return this;
+        } else if (this.fragment) {
+            return new Path(this.directory);
+        } else {
+            return new Path(stdPath.dirname(this.directory) + this.sep);
+        }
+    }
+
+    /**
+     * Return path for th root directory for the drive this path is on.
+     */
+    root() {
+        let last = null;
+        let current = this.full;
+        while (current !== last) {
+            last = current;
+            current = stdPath.dirname(current);
+        }
+
+        return new Path(current);
+    }
+
+    /**
+     * Create an empty file at the given path if it doesn't already exist.
+     */
+    createFile() {
+        touch.sync(this.full);
+    }
+
+    /**
+     * Create directories for the file this path points to, or do nothing
+     * if they already exist.
+     */
+    createDirectories() {
+        try {
+            mkdirp.sync(this.directory);
+        } catch (err) {
+            if (err.code !== 'ENOENT') {
+                throw err;
+            }
+        }
+    }
+
+    matchingPaths(caseSensitive=null) {
+        let absoluteDir = absolutify(this.directory);
+        let filenames = null;
+
+        try {
+            filenames = fs.readdirSync(absoluteDir);
+        } catch (err) {
+            return []; // TODO: Catch permissions error and display a message.
+        }
+
+        if (this.fragment) {
+            if (caseSensitive === null) {
+                caseSensitive = this.hasCaseSensitiveFragment();
+            }
+
+            filenames = [
+                for (fn of filenames)
+                if (matchFragment(this.fragment, fn, caseSensitive))
+                fn
+            ];
```

Comment:
I'm pretty sure that array comprehensions got removed from es2015. At least, I don't see them in any Babel stage. Might be wise to replace this with `filenames = filenames.filter(fn => matchFragment(this.fragment, fn, caseSensitive))`.